testcases/OpTestKexec.py

```diff
@@ -27,143 +27,387 @@
+        reboot. Reboot command with kexec kernel loaded will lead to clean
+        kexec.
+
+        Args:
+            exec_opt (bool, optional): If True, generates a kexec command for
+                                       immediate execution. If False (default),
+                                       generates a kexec command for system
+                                       reboot.
+
+        Returns:
+            str: The generated kexec command.
+        """
+
+        kexec_cmd = None
+
+        if exec_opt:
+            kexec_cmd = "kexec -e"
+        else:
+            kexec_cmd = "reboot"
+
+        return kexec_cmd
+
+    def is_kexec_kernel_loaded(self):
+        """
+        Checks if a kexec kernel is currently loaded.
+
+        Returns:
+            bool: True if a kexec kernel is loaded, False if not loaded, and
+                  None if the status cannot be determined.
         """
+
         try:
-            raw_pty = self.cv_SYSTEM.console.get_console()
-            raw_pty.sendline(cmd)
-            raw_pty.expect("login:", timeout=600)
-            log.info("{} passed(SUCCESS)".format(cmd))
+            kexec_loaded = self.c.run_command("cat /sys/kernel/kexec_loaded")
+            if kexec_loaded[0] == "1":
+                return True
         except CommandFailed:
-            log.info("{} failed with exception.".format(cmd))
+            log.error("Failed to get kexec loaded status.")
+
+        return False
 
-    def kexec_load_unload(self, arg):
+    def execute_kexec_cmd(self, cmd, raw_pty_console=False):
         """
-        This function will test the kexec load or unload based on arg
+        Executes a command using either the standard shell or a raw PTY
+        console.
+
+        Args:
+            cmd (str): The command to be executed.
+            raw_pty_console (bool, optional): If True, uses a raw PTY console
+                                              for execution.
+
+        Returns:
+            bool: True if the command runs successfully, False if there's an
+                  error.
         """
+
         try:
-            if arg is "l":
-                self.c.run_command("kexec -l /boot/{} --initrd /boot/{} --append=\"`cat /proc/cmdline`\"".format(self.kernel_image,self.initrd_image))
-            elif arg is "u":
-                self.c.run_command("kexec -u")
+            if raw_pty_console:
+                raw_pty = self.cv_SYSTEM.console.get_console()
+                raw_pty.sendline(cmd)
+                raw_pty.expect("login:", timeout=600)
+                # Brings back the system to normal state
+                # Below function reboots the system which is OK
+                # for the test other then kexec in loop
+                # TODO need to find a better way bring the system in
+                # consistent withough reboot it.
+                self.cv_SYSTEM.set_state(OpSystemState.UNKNOWN)
+                self.cv_SYSTEM.goto_state(OpSystemState.OS)
```

Comment:
Though these function resets the system state and gets you the logged in os prompt for next kexec command, but adds up extra reboot cycles. Can you try closing the raw_pty and open and get a new console, this will allow you to get new raw_pty with prompt reset to run next command.

Reply:
So currently there are five test cases for kexec:

    1. kexec load and unload
    2. kexec load and exec into the new kernel
    3. kexec using the force option
    4. kexec using single step
    5. kexec in a loop
    
So for test 1 we don't do system reset because raw_pty is not used. Now from test 2, 3, and 4 I think resetting the system is good because for subsequent test the system state will be consistent.

Your input will be helpful for test 5 where system should not reboot if kexec need to be performed in loop.

As per offline discussion I will try closing the console and get it again for test 5 and see if that helps.

Thanks for the input @abdhaleegit 

- Sourabh

Reply:
@abdhaleegit Closing the console connection worked nicely for kexec in loop test case.

So now for the test 1 "kexec load and unload" no system reset is done
for test 2 - 4 system is reset which I think should be done to avoid kexec on kexec across test
and for test 5 kexec in loop test, console is closed for system login for next kexec in loop.

Thanks for suggestion.

Reply:
@sourabhjains this was the commit I was talking about https://github.com/open-power/op-test/pull/739  with out this even closing and taking back the console  did not set the prompt for next user

Reply:
Oh Ok. But now closing and taking back console is working fine.